### ISSUE TITLE
Custom CNN-based handwriting-to-answer verifier (replacing generic OCR)

### ISSUE THREAD

**Issue by hoj201** [hoj201/grader_bot#81]
## Context

GraderBot currently grades handwritten answers by running one of three generic
OCR backends (Mathpix, EasyOCR, Google Cloud Vision — `graderbot/answer_reader.py`)
over an answer-box crop, then string-comparing the result to the stored answer
(`grading.py:grade_response`). This is mediocre because none of these models
know the worksheet's math context: Mathpix (the default) is tuned for
college-level typeset math and misreads sloppy digits (issue #70: "9"→"G",
"14"→"1 h") and hand-drawn fraction bars (issue #39, patched after the fact
by `_iter_fraction_reinterpretations`). A prior attempt to fix this with a
pretrained CTC handwriting model (`handwriting-ctc-match` branch, issue #73,
using `pylaia-iam`) stalled: that model was trained on English prose, so it
barely represents digits and has essentially no exposure to `/`, and it
couldn't discriminate the "1" vs "/" confusion it was meant to fix. That spike
is paused, not wired into grading, and has no real-data validation.

The instinct to train a custom model is right, but the framing needs one
correction: **no bounding-box detector is needed.** Answer-box positions are
already known deterministically (`worksheet_boxes.py` extracts them once
from the vector-PDF template; `registration.py` homography-warps every scan
onto that same fixed geometry). The actual gap is recognition/verification of
a crop whose location is already known, and — critically — whose **correct
answer is already known too** (it's in the answer key). That turns this from
an open-vocabulary OCR problem into a much smaller **closed-set verification**
problem: given a crop and a short list of candidate strings (the correct
answer plus a few plausible OCR-confusable near-misses), score which
candidate the crop actually shows. This directly reuses the CTC-scoring idea
from issue #73 (which was structurally sound) while fixing its actual failure
mode — a model trained from scratch on a ~15-symbol domain vocabulary (digits,
`.`, `-`, `/`, space) instead of a borrowed 81-symbol English-prose model.

Design decisions:
- **Verification against candidates**, not open-vocabulary recognition — reuses existing box localization, no bbox detector.
- **Real validation data via a dedicated labeling pass** (not synthetic-only), to catch the same synthetic/real domain gap that sank the PyLaia spike.
- **Start fresh** rather than resurrecting the `handwriting-ctc-match` branch's code — but keep its two hard-won lessons: (1) domain-appropriate training data is the whole game, and (2) raw CTC probabilities are small and non-thresholdable by construction, so scoring must be relative (candidate vs. candidate), never an absolute cutoff.

## Model design

A small CRNN (few conv+pool blocks → 1-2 layer bidirectional GRU/LSTM →
linear + log-softmax), trained from scratch with CTC loss on a tiny vocabulary
(`0-9 . - / <space> <blank>`, ~15 symbols vs. PyLaia's 81) — well under 1M
parameters, CPU-trainable. At inference, reimplement the CTC forward
algorithm (log-space alpha recursion) to compute `log P(candidate | crop)`
for each of ~3-6 candidates, then **softmax-normalize across the candidate
set** rather than thresholding absolutely — this is the direct fix for issue
#73's "scores come back small" lesson.

Rejected alternatives: a Siamese/contrastive crop-vs-rendered-candidate
encoder (just relocates the domain-mismatch risk into embedding space without
the benefit of an already-validated relative-scoring mechanism); a fixed-slot
per-character classifier (breaks on variable-length answers — fractions,
decimals, negatives, mixed numbers).

## Data pipeline

**Synthetic (bulk of training data):** new `graderbot/answer_glyph_synth.py`,
a fast PIL-based box-scale renderer (`worksheet_synth.fill_worksheet` is too
slow for large-scale generation — full LaTeX/PDF compile per call). Reuses
`worksheet_synth.py`'s existing centering/fraction-bar drawing logic
(`_draw_text_centered_in`, `_draw_answer`'s fraction handling, `_FRAC_RE`).
Add more handwriting-style fonts beyond the current single
`fonts/HomemadeApple-Regular.ttf`, plus stroke-width/slant/spacing/baseline
jitter layered on the existing `perspective_skew_image`/`add_image_noise`.

**Deliberate near-misses:** new `graderbot/response_candidates.py`,
`generate_candidates(answer, rng) -> List[str]` — digit-confusion swaps
(cataloguing issue #70's known misreads), fraction-bar reinterpretations
(mirroring `_iter_fraction_reinterpretations`'s shape but generating strings),
unreduced-fraction variants (stresses the existing "must simplify" rule in
`grade_response`), sign/decimal-shift variants. Shared by both training
(hard negatives) and runtime (building the actual candidate list).

**Real labeling pass:** new `HANDWRITING_LABEL` table in `storage.py`
(mirrors `NAME_IMAGES`/`MATHPIX_CALL` conventions — content-addressed S3 key,
`source_mathpix_call_id` linking back to existing `MATHPIX_CALL` rows as a
seed). A CLI script (`scripts/label_handwriting.py`) rather than a Streamlit
page — this is a one-off solo labeling pass, not a repeated multi-user flow.
Real data is used first as a **held-out validation set** (explicitly
measuring the synthetic-vs-real accuracy gap — the metric that would have
caught the PyLaia failure sooner) and only folded into fine-tuning once
there's enough of it (rough threshold: >150-200 labeled real crops).

## Training/eval infra

New top-level `training/` directory, isolated the same way `easyocr_service/`
isolates torch: its own `requirements.txt`, **never added to
`pyproject.toml`**, so neither Intel-Mac local dev nor the fly.io deploy image
ever needs torch. Contains `model.py`, `dataset.py`, `train.py` (torch +
`ctc_loss`), `export_onnx.py` (`torch.onnx.export` + a numerical-parity check
against sample crops, mirroring the paused branch's verified ~2e-5 max-diff
check), `eval.py` (per-answer-type accuracy, confusion matrix over
`response_candidates`' near-miss categories, and the synthetic-vs-real gap —
same spirit as `name_classifier.loo_cross_validate`). Training runs offline
(local or Modal, already used for `easyocr_service`), never in CI.

The pure-numpy CTC forward algorithm used at *inference* time (no torch)
lives in `graderbot/response_scorer.py` and gets a normal pytest under
`tests/`, validated against brute-force enumeration — same technique the
paused branch already proved out in `tests/test_handwriting_match.py`.

## Integration

New `ResponseScorer` protocol in `graderbot/response_scorer.py`, modeled on
`NameReader` (`graderbot/name_reader.py:60-66`) rather than `AnswerReader`,
because scoring needs the candidate set, which only `grade_hw` can build (it
alone has `answer_key[qid]`):

```python
class ResponseScorer(Protocol):
    def score(self, image: np.ndarray, box: Box, candidates: List[str]) -> ScoredResponse: ...
```

`CnnResponseScorer` lazy-loads the ONNX session + vocab, same pattern as
`ClassifierNameReader.from_classroom`'s lazy `name_classifier` import.

Wiring, parallel to the existing `answer_reader` path at every level, always
optional and defaulting to today's behavior:
- `grade_hw` (`grading.py:17`) gains an optional `response_scorer` param. When set: build `candidates = [answer] + response_candidates.generate_candidates(answer)`, call `response_scorer.score(...)`, set `response = scored.best`, then still run the existing `grade_response(response, answer)` unchanged — fraction-simplification and all comparison rules stay centralized in one place.
- `scan_grader._grade_batch` (`scan_grader.py:102`) threads `response_scorer` through the same way `answer_reader` is threaded today.
- The existing blank-box short-circuit (`grading.py:51`, `imaging.is_blank`) runs before either path — the verifier never needs a "blank" class, avoiding the gap `ClassifierNameReader` had to patch (issue #62).
- `app.py`'s Grade tab (`render_grade`, `~app.py:873`): add a 4th entry to the answer-reader dropdown (`~app.py:908`), e.g. "CNN verifier (experimental)", constructing `CnnResponseScorer()` the same way `EasyOcrAnswerReader`/`GoogleVisionAnswerReader` are constructed today (`~app.py:967-982`).

## Deployment

`onnxruntime` becomes a direct `pyproject.toml` dependency (unlike EasyOCR's
torch, it ships real wheels for both Intel macOS and Linux/fly.io — the
paused branch already confirmed this at `onnxruntime~=1.23.2`). This lets
`CnnResponseScorer` be in-process like `GoogleVisionAnswerReader`, no sidecar,
no Dockerfile/docker-compose changes. Model artifact
(`models/response_scorer/weights.onnx` + `vocab.json`, ~tens of MB) committed
to the repo. **Fallback**, only if Phase 0 finds no viable Intel-macOS
`onnxruntime` wheel: fall back to an `easyocr_service`-style sidecar; the
`ResponseScorer` protocol is unaffected either way.

## Phasing (each phase independently shippable/testable; new pytest per CLAUDE.md)

0. Confirm `onnxruntime` wheel availability (Intel macOS + Linux/py3.13) — go/no-go on deployment shape.
1. `graderbot/response_candidates.py` — pure logic, `tests/test_response_candidates.py`. Standalone value even before any model exists.
2. `graderbot/answer_glyph_synth.py` — fast synthetic renderer, multi-font + jitter. `tests/test_answer_glyph_synth.py` (deterministic given seed, non-blank via `imaging.ink_fraction`).
3. `HANDWRITING_LABEL` table + storage functions (`tests/test_storage.py` additions, moto-mocked S3) + `scripts/label_handwriting.py` CLI.
4. `training/` — model, dataset, train, export_onnx, eval; pure-numpy CTC scorer lands in `graderbot/response_scorer.py` with brute-force-validated tests.
5. `ResponseScorer`/`CnnResponseScorer` in `graderbot/response_scorer.py`. `tests/test_response_scorer.py` with a faked `onnxruntime.InferenceSession` — ships/tests before Phase 4 produces a *good* model.
6. Wire into `grade_hw`, `_grade_batch`, `app.py` Grade tab dropdown. Extend existing `grade_hw`/`_grade_batch` tests with a fake `ResponseScorer` case; extend `tests/test_app.py` (Streamlit `AppTest`) for the new dropdown option.
7. Real-data eval loop (once Phase 3 produces enough labels): run `training/eval.py`, compare synthetic-vs-real accuracy gap, decide on fine-tuning. Keep the Grade tab option labeled "(experimental)" until real-eval accuracy is competitive with Mathpix. Iterative, not one-time.

## Verification

- `poetry run pytest` after each phase (per CLAUDE.md) — Phases 1, 2, 3, 5, 6 each add real pytest coverage without needing a trained model.
- Phase 4's `training/eval.py` run by hand (or via Modal) reports per-answer-type accuracy and the synthetic-vs-real gap — the number that determines whether Phase 7 is "ready" or "needs more real data/retraining."
- End-to-end: once Phase 6 lands, run a real scanned worksheet through the Grade tab with "CNN verifier (experimental)" selected and compare its per-question results/confidence against the same run with Mathpix, using the existing marked-up-PDF output to eyeball agreement/disagreement.

## Related

Builds on the lessons from #73 (paused pretrained-CTC spike) and addresses
the misread patterns documented in #70 and #39. Does *not* build on the
`handwriting-ctc-match` branch's code — starting fresh with a domain-trained
model instead of a borrowed English-prose one.

🤖 _Posted by Claude Code on behalf of @hoj201_
